net: openthread: add Link Metrics API

lmaciejonczyk opened the pull request:
This commit implements the OpenThread APIs to configure Enhanced-ACK Based Probing in radio for a specific Initiator. This is needed for Link Metrics functionality.

edmont commented on subsys/net/l2/openthread/Kconfig.features:
```diff
@@ -182,3 +182,6 @@ config OPENTHREAD_OTNS
 config OPENTHREAD_FULL_LOGS
 	bool "Enable OpenThread full logs"
+
+config OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
```
```suggestion
config OPENTHREAD_MLE_LINK_METRICS_ENABLE
```
Same way as with other configurations.

lmaciejonczyk committed:
net: openthread: add Link Metrics API

This commit implements the OpenThread APIs to configure Enhanced-ACK
Based Probing in radio for a specific Initiator. This is needed for
Link Metrics functionality.

Signed-off-by: Lukasz Maciejonczyk <lukasz.maciejonczyk@nordicsemi.no>

edmont commented on include/net/ieee802154_radio.h:
```diff
@@ -208,6 +213,16 @@ struct ieee802154_config {
+
+		/** ``IEEE802154_CONFIG_ENH_ACK_PROBING`` */
+		struct {
+			bool lqi : 1;
```
Thread specification requires only 1 or 2 of among (LQI, LM, RSSI)  to be selected, and order is to be kept. This structure doesn't warranty those requirements. However, OpenThread implementation neither, so maybe such improvement can be added in the future.

edmont commented on include/net/ieee802154_radio.h:
```diff
@@ -208,6 +213,16 @@ struct ieee802154_config {
+			bool rssi : 1;
+			uint16_t short_addr;
+			const uint8_t *ext_addr;
+
```
nit: unnecessary blank line.

rlubos commented on include/net/ieee802154_radio.h:
```diff
@@ -162,6 +162,11 @@ enum ieee802154_config_type {
+	/** Enable/disable or update Enhanced-ACK Based Probing in radio
+	 *  for a specific Initiator.
+	 */
+	IEEE802154_CONFIG_ENH_ACK_PROBING,
```
It's a bit unclear to me, OpenThread enables probing, and what then? Will this trigger some callbacks from the radio side?

edmont replied:
From this moment the device should include the configured link metrics values in every Enh-ACK generated for this neighbor.

rlubos replied:
Ah, so that's how it works, thanks for clarifying.

galak commented:
@lmaciejonczyk can you rebase to address the merge conflicts.

aurel32 commented:
On the `openthread` module, there is zephyr/CMakeLists.txt:

```
if(CONFIG_OPENTHREAD_LINK_METRICS)
  set(OT_LINK_METRICS ON CACHE BOOL "Enable Link Metrics feature for Thread 1.2")
endif()
```

`OT_LINK_METRICS` basically sets `OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE=1`, so it seems to be a different way to enable Link Metrics. However there is nothing that allows to enable `CONFIG_OPENTHREAD_LINK_METRICS` on the Zephyr side.

lmaciejonczyk commented:
> On the `openthread` module, there is zephyr/CMakeLists.txt:
> 
> ```
> if(CONFIG_OPENTHREAD_LINK_METRICS)
>   set(OT_LINK_METRICS ON CACHE BOOL "Enable Link Metrics feature for Thread 1.2")
> endif()
> ```
> 
> `OT_LINK_METRICS` basically sets `OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE=1`, so it seems to be a different way to enable Link Metrics. However there is nothing that allows to enable `CONFIG_OPENTHREAD_LINK_METRICS` on the Zephyr side.

Good catch, thanks!
Will be fixed by #35035
